eve: Create a single terraform worker (with ability to use snapshot)

TeddyAndrieux opened the pull request:
**Component**:

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

#2842 

**Summary**:

Instead of creating several different worker with basically a copy/paste
for each different kind of environment we need using terraform, create a
single one.
Some update:
- Create 2 networks for Control and Workload plane
- Create Ports with static Ips for every machines
- Add a terraform variable to spawn from snapshot (and then restore a
  running MetalK8s cluster, will be used mainly for lifecycle tests)

**Acceptance criteria**:

- [x] Multi-node installation
- [x] RHEL single-node
- [ ] ~Bootstrap restore~ (#2886)
- [ ] ~Upgrade using snapshot~ (#2889)

---

TODO:

- [x] Handle RHEL in this terraform worker
- [x] Remove openstack rhel specific worker
- [x] Handle Cypress in this terraform worker
- [x] Remove openstack cypress specific worker
- [x] Use new terraform worker in CI (update `eve/main.yaml`)
- [x] Create a snapshot for single-node 2.5.1
- [x] Use snapshot in single-node upgrade for released versions

---

Fixes: #2842

TeddyAndrieux committed:
eve: Create a single terraform worker

Instead of creating several different worker with basically a copy/paste
for each different kind of environment we need using terraform, create a
single one. This commit add support for CentOs 7 multi-node deployment
Some update:
- Create 2 networks for Control and Workload plane
- Create Ports with static Ips for every machines
- Rename `nodeX` by `node-X` everywhere
- Add a terraform variable to spawn from snapshot (and then restore a
  running MetalK8s cluster, will be used mainly for lifecycle tests)

Sees: #2842

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/GH-2842-use-snapshot-in-ci` into
`development/2.6`
* [w/2.7/improvement/GH-2842-use-snapshot-in-ci](https://github.com/scality/metalk8s/compare/development/2.7...w/2.7/improvement/GH-2842-use-snapshot-in-ci) will be merged into `development/2.7` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @nicolast

* @lucieleonard

* @alexandre-allard-scality

* @monpote

* @slaperche-scality

* @teddyandrieux

* @cmonfort

* @thomasdanan

* @carlito-scality

* @jbertran

* @sayf-eddine-scality

* @gdemonet

* @nootal

TeddyAndrieux committed:
eve: Add rhel support in `openstack-terraform` worker

Add support for RedHat spawn using single terraform
`openstack-terraform` worker

bert-e commented:
# History mismatch



Merge commit #a8c4b8976be19f926c3538e5324e2739d1ef66ee on the integration branch
`w/2.7/improvement/GH-2842-use-snapshot-in-ci` is merging a branch which is neither the current
branch `improvement/GH-2842-use-snapshot-in-ci` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/GH-2842-use-snapshot-in-ci` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux committed:
eve: Add cypress support in `openstack-terraform` worker

Add support for Cypress spawn using single terraform
`openstack-terraform` worker (it basically install Cypress requirements
on the Bastion VM, so that we can run Cypress test from Bastion)

TeddyAndrieux committed:
eve: Use snapshot for upgrade of promoted versions in CI

In post-merge in the CI we want to test upgrade from promoted versions
(last minor release and last patch release), but it does not make sense
to install this version on every tests, instead lets use snapshots.
Add some logic in eve so that we can easily add new upgrades tests from
other snapshots.

sees: #2842

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/GH-2842-use-snapshot-in-ci` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `improvement/GH-2842-use-snapshot-in-ci`
* Rebase `improvement/GH-2842-use-snapshot-in-ci` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# History mismatch



Merge commit #a8c4b8976be19f926c3538e5324e2739d1ef66ee on the integration branch
`w/2.7/improvement/GH-2842-use-snapshot-in-ci` is merging a branch which is neither the current
branch `improvement/GH-2842-use-snapshot-in-ci` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/GH-2842-use-snapshot-in-ci` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.7/improvement/GH-2842-use-snapshot-in-ci` with contents from `improvement/GH-2842-use-snapshot-in-ci`
and `development/2.7`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.7/improvement/GH-2842-use-snapshot-in-ci origin/development/2.7
 $ git merge origin/improvement/GH-2842-use-snapshot-in-ci
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.7/improvement/GH-2842-use-snapshot-in-ci
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @gdemonet

* @alexandre-allard-scality

* @jbertran

* @sayf-eddine-scality

* @teddyandrieux

* @nicolast

* @lucieleonard

* @nootal

* @monpote

* @cmonfort

* @slaperche-scality

* @tcarmet

* @carlito-scality

* @thomasdanan

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

alexandre-allard commented on tests/conftest.py:
```diff
@@ -1,4 +1,5 @@
 import json
+import os.path
```
any reason for this import ?

TeddyAndrieux replied:
Good catch

alexandre-allard commented on eve/workers/openstack-terraform/terraform/scripts/network.sh:
```diff
@@ -0,0 +1,22 @@
+
+IFACE=$1
+
+MAC_ADDR="$(cat "/sys/class/net/$IFACE/address")"
```
Useless use of cat & subshell:
```suggestion
read -r MAC_ADDR < "/sys/class/net/$IFACE/address"
```

TeddyAndrieux replied:
Right :+1:

bert-e commented:
# History mismatch



Merge commit #ff60ad9e88da0eed8ac4b545165462fab0f9cd12 on the integration branch
`w/2.7/improvement/GH-2842-use-snapshot-in-ci` is merging a branch which is neither the current
branch `improvement/GH-2842-use-snapshot-in-ci` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/GH-2842-use-snapshot-in-ci` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented on eve/workers/openstack-terraform/terraform/nodes.tf:
```diff
@@ -0,0 +1,314 @@
+  # Register RHSM if OS = rhel
+  provisioner "remote-exec" {
+    inline = [
+      "if [ '${var.os}' = 'rhel-7' ]; then sudo chmod +x scripts/rhsm-register.sh; fi",
```
```suggestion
      "[ '${var.os}' = 'rhel-7' ] && sudo chmod +x scripts/rhsm-register.sh",
```

TeddyAndrieux replied:
Yes but no because at the end the retcode will be `1` if we are not on rhel-7

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.7/improvement/GH-2842-use-snapshot-in-ci` with contents from `improvement/GH-2842-use-snapshot-in-ci`
and `development/2.7`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.7/improvement/GH-2842-use-snapshot-in-ci origin/development/2.7
 $ git merge origin/improvement/GH-2842-use-snapshot-in-ci
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.7/improvement/GH-2842-use-snapshot-in-ci
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #2d6e269a8feafd3dac225853a7ae1e56e8dcab92 on the integration branch
`w/2.7/improvement/GH-2842-use-snapshot-in-ci` is merging a branch which is neither the current
branch `improvement/GH-2842-use-snapshot-in-ci` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/GH-2842-use-snapshot-in-ci` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.7/improvement/GH-2842-use-snapshot-in-ci` with contents from `improvement/GH-2842-use-snapshot-in-ci`
and `development/2.7`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.7/improvement/GH-2842-use-snapshot-in-ci origin/development/2.7
 $ git merge origin/improvement/GH-2842-use-snapshot-in-ci
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.7/improvement/GH-2842-use-snapshot-in-ci
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #d4dbc51d50e19fab09fbd985800982842c82d078 on the integration branch
`w/2.7/improvement/GH-2842-use-snapshot-in-ci` is merging a branch which is neither the current
branch `improvement/GH-2842-use-snapshot-in-ci` nor the development branch
`development/2.7`.

It is likely due to a rebase of the branch `improvement/GH-2842-use-snapshot-in-ci` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.7/improvement/GH-2842-use-snapshot-in-ci` with contents from `improvement/GH-2842-use-snapshot-in-ci`
and `development/2.7`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.7/improvement/GH-2842-use-snapshot-in-ci origin/development/2.7
 $ git merge origin/improvement/GH-2842-use-snapshot-in-ci
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.7/improvement/GH-2842-use-snapshot-in-ci
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

alexandre-allard-scality commented:
/approve

bert-e commented:
# Not Author



I'm afraid I cannot do that, @alexandre-allard-scality:

Only the author of the pull request @teddyandrieux can use this command.

Please **delete** the corresponding comment so I can move on.

TeddyAndrieux commented:
/approve

alexandre-allard reviewed (approved):
Tested with a snapshot, it worked well. LGTM.

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.6`

* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`



Please check the status of the associated issue GH-2842.

Goodbye teddyandrieux.
